### PR TITLE
INFRA-427: Add a new deployment type to accommodate the new Ozone packaging

### PR DESCRIPTION
**Pull request by enyachoke**
This PR handles changes needed for https://mekomsolutions.atlassian.net/browse/INFRA-427 and https://mekomsolutions.atlassian.net/browse/INFRA-428.  I still need to investigate an Issue with a failing test but we can start reviewing the code. The name of the project type is not final so suggestions are welcome.

### CODE REVIEW

**Comment by rbuisson** [comment]
Thanks @enyachoke  for this good piece of work.

With this new deployment type, are we allowed to provide anything as Maven artifact anymore? Or has this been deprecated?

**Comment by enyachoke** [comment]
> Thanks @enyachoke for this good piece of work.
> 
> With this new deployment type, are we allowed to provide anything as Maven artifact anymore? Or has this been deprecated?
You can still provide a Maven artifact

**Review by rbuisson — COMMENTED** [review]
Sorry, I've had started to review and got interrupted.
Let me publish what I had commented.

File: node-scripts/src/pipeline3/scripts.js [hunk 1]
```diff
@@ -11,25 +11,35 @@ const heredoc_3 = cst.HEREDOC_3;
 const utils = require("../utils/utils");
 const model = require("../utils/model");
 const createEnvVarFile = function(instanceDef, dockerComposePath) {
```
Comment (rbuisson) [inline node-scripts/src/pipeline3/scripts.js:13]: I don't think this function should be Docker specific, right?
Please rename to something more generic.

File: node-scripts/spec/pipeline3/impl/dockerComposeGenericRegacy.spec.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    type: "dev",
+    group: "tlc",
+    deployment: {
+      hostDir: "/var/docker-volumes/",
+      type: "dockerComposeGenericMaven",
+      composePlugin: true,
+      dockerComposeFiles: ["docker-compose.yml", "docker-compose-2.yml"],
+      envFiles: ["env-file-1", "env-file-2"],
+      value: {
+        mavenProject: {
```
Comment (rbuisson) [inline node-scripts/spec/pipeline3/impl/dockerComposeGenericRegacy.spec.js:28]: Let's remove support for `mavenProject` as part of this new type.

File: node-scripts/spec/pipeline3/impl/dockerComposeGeneric.spec.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  ));
+
+  var instanceDef = {
+    uuid: "336af1ee-90a1-4d1b-baaf-db12c84deec0",
+    name: "cambodia1",
+    type: "dev",
+    group: "tlc",
+    deployment: {
+      hostDir: "/var/docker-volumes/",
+      workDir: "/var/docker-volumes/artifacts/run/docker",
```
Comment (rbuisson) [inline node-scripts/spec/pipeline3/impl/dockerComposeGeneric.spec.js:23]: Let's move this key to `value` and rename it to `projectPath`

**Review by rbuisson — CHANGES_REQUESTED** [review]

File: node-scripts/spec/pipeline3/impl/dockerComposeGenericRegacy.spec.js (hunk 2 above)
Comment (rbuisson) [reply node-scripts/spec/pipeline3/impl/dockerComposeGenericRegacy.spec.js:28]: Please rename the file, if it's still needed, `dockerComposeGenericLegacy.spec.js`

File: node-scripts/src/utils/config.js [hunk 4]
```diff
@@ -382,7 +382,8 @@ module.exports = {
       "dockerCompose",
       "dockerComposeGit",
       "dockerComposeMaven",
-      "dockerComposeGenericMaven"
+      "dockerComposeGenericMaven",
+      "dockerComposeProjectDir"
```
Comment (enyachoke) [inline node-scripts/src/utils/config.js:386]: Rename to dockerComposeFromArtifact.

File: node-scripts/src/utils/model.js [hunk 5]
```diff
@@ -68,12 +68,19 @@ class DockerComposeMavenDeployment {
...
 class DockerComposeGenericMavenDeployment {
-  constructor(mavenProject, services) {
+  constructor(mavenProject, services, mavenUrl) {
     this.mavenProject = mavenProject;
     this.services = services;
     this.mavenUrl = this.mavenUrl;
   }
 }
+
+class DockerComposeGenericDeployment {
```
Comment (enyachoke) [inline node-scripts/src/utils/model.js:78]: Add some JS doc describing the type.
```
A type which assumes that the Docker Compose project will be installed through the processing of the `artifacts` section.
```

Comment (rbuisson) [reply node-scripts/src/utils/model.js]: @enyachoke , did you drop this in your latest commit?

**Review by rbuisson — CHANGES_REQUESTED** [review]
It seems that you've dropped this addition:
> A type which assumes that the Docker Compose project will be installed through the processing of the `artifacts` section.

File: node-scripts/src/instance-event/validator.js [hunk 6]
```diff
@@ -248,6 +247,18 @@ module.exports = {
...
   },
+  validateDockerComposeFromArtifactDeploymentConfigValue: function(value) {
+    if (
+      JSON.stringify(Object.keys(value).sort()) >=
+      JSON.stringify(
+        Object.keys(new model.DockerComposeFromArtifactDeployment()).sort()
+      )
+    ) {
+      throw new Error(
+        "The Docker compose deployment value should be provided as an instance of 'DockerComposeFromArtifactDeployment'."
```
Comment (rbuisson) [inline node-scripts/src/instance-event/validator.js:258]: Could you replace by:
> The 'DockerComposeFromArtifact' deployment value should be provided as an instance of...

**Review by rbuisson — APPROVED** [review]
Tried it and works good.

**Pull request merged**